File: Tatelier.Common/SongSelect/ItemControl.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Tatelier.Common.SongSelect
{
	/// <summary>
	/// 選択項目管理クラス
	/// </summary>
	public class ItemControl
	{
		public static ItemControl Create(string folder, CategoryControl categoryControl, out string[] error)
		{
			var instance = new ItemControl(folder, categoryControl);

			error = null;

			return instance;
		}

		/// <summary>
		/// 項目リスト
		/// </summary>
		public List<IItem> ItemList = new List<IItem>();

		/// <summary>
		/// 譜面フォルダ
		/// </summary>
		public string ScoreFolder;

		/// <summary>
		/// コンストラクタ
		/// </summary>
		/// <param name="scorefolder">譜面フォルダ</param>
		/// <param name="categoryControl">カテゴリ管理</param>
		public ItemControl(string scorefolder, CategoryControl categoryControl)
		{
			ScoreFolder = scorefolder;
			ItemList.AddRange(categoryControl.CategoryMap.Values.Distinct());

			var files = Directory.EnumerateFiles(scorefolder, "*", SearchOption.AllDirectories).Where(v =>
			{
				var extension = Path.GetExtension(v);
				switch (extension)
				{
					case ".tja":
					case ".pts":
					case ".tlscore":
						return true;
					default:
						return false;
				}
			});

			foreach (var item in files.Select(v => v.Replace(scorefolder + "\\", "")) ?? new string[0])
			{
				_ = ScoreOverview.Create(scorefolder, item, categoryControl);
			}
		}
	}
}

File: Tatelier.Nucleus/SongSelect/ItemControl.cpp
#include "ItemControl.h"

#include "../Error.h"

#include "../Utility.h"
#include "ScoreOverview.h"

namespace SongSelect {
	ItemControl* ItemControl::Create(const std::string& folder, std::shared_ptr<CategoryControl> categoryControl)
	{
		err.clear();

		return new ItemControl(folder, categoryControl);
	}
	std::shared_ptr<std::vector<std::shared_ptr<Item>>> ItemControl::GetItemList()
	{
		return itemList;
	}
	const std::string& ItemControl::GetScoreFolder()
	{
		return scoreFolder;
	}
	ItemControl::ItemControl(const std::string& scoreFolder, std::shared_ptr<CategoryControl> categoryControl)
	{

		this->scoreFolder = scoreFolder;
		auto& categoryMap = categoryControl->GetCategoryMap();

		itemList = std::shared_ptr<std::vector<std::shared_ptr<Item>>>(new std::vector<std::shared_ptr<Item>>());

		for (auto category : categoryMap) {
			std::shared_ptr<Item> pointer = std::static_pointer_cast<Item>(category.second);
			if (std::find(itemList->begin(), itemList->end(), pointer) == itemList->end()) {
				itemList->push_back(pointer);
			}
		}

		std::filesystem::path scoreRootFolder;

		scoreRootFolder /= scoreFolder;
		scoreRootFolder /= "Root";

		std::vector<std::string> files;
		if (ttleGetFileNamesRecursive(scoreRootFolder.string(), &files, [](std::string a) {
			return ttleEndWithIndex(a, ".tja|.tlscore") > 0;
			})) {
			for (auto& str : files) {
				ScoreOverview::CreateAndRegist(scoreFolder, str, categoryControl);
			}
		}
	}
}